### PR TITLE
Refactor HLRC RequestConverters parameters to be more explicit

### LINKED ISSUE #39666: High level REST client parameters look wrong due indirectness and side effects
Using the high-level request client, this is how code to build a request with parameters appears today:

```java
        final Request request = new Request(HttpPut.METHOD_NAME, endpoint);
        final RequestConverters.Params parameters = new RequestConverters.Params(request);
        parameters.withWaitForActiveShards(putFollowRequest.waitForActiveShards());
        request.setEntity(createEntity(putFollowRequest, REQUEST_BODY_CONTENT_TYPE));
```

This looks wrong because `parameters` appears to be unused. That is, we create a `RequestConverter.Params` object, mutate it, and then it appears nothing happens to it. This code is not broken though, because what is happening behind the scenes is that `request` is being mutated when methods on `parameters` are invoked. This indirectness and methods having side effects on another object lead to code that does not appear correct at first glance, and is more confusing than it should be; it looks wrong. Instead, I propose an alternative like:

```java
        final Request request = new Request(HttpPut.METHOD_NAME, endpoint);
        RequestConverters.Params parameters = new RequestConverters.Params();
        parameters.withWaitForActiveShards(putFollowRequest.waitForActiveShards());
        request.setParameters(parameters);
        request.setEntity(createEntity(putFollowRequest, REQUEST_BODY_CONTENT_TYPE));
```

This makes apparent that the `parameters` are actually used, and that they have an effect on the `request` object through explicit invocation. I believe this makes it more transparent what is occurring.

### PR DESCRIPTION
**Pull request by ojasgulati**
The existing `RequestConverters.Params` is confusing, because it wraps
an underlying request object and mutations of the `Params` object
actually mutate the `Request` that was used in the construction of the
`Params`.

This leads to a situation where we create a `RequestConverter.Params`
object, mutate it, and then it appears nothing happens to it - it
appears to be unused. What happens behind the scenes is that the Request
object is mutated when methods on `Params` are invoked. This results in
unclear, confusing code where mutating one object changes another with
no obvious connection.

This commit refactors `RequestConverters.Params` to be a simple helper
class to produce a `Map` which must be passed explicitly to a Request
object. This makes it apparent that the `Params` are actually used, and
that they have an effect on the `request` object explicit and easier to
understand.

Resolves https://github.com/elastic/elasticsearch/issues/39666

### CODE REVIEW

**Comment by tvernum** [comment]
@ojasgulati 
You seem to have mixed some watcher changes in with your HL Rest Client changes.
Can you split them into separate PRs so that they can be reviewed independently.

Thanks.

**Comment by tvernum** [comment]
@elasticmachine Test this please

**Comment by ojasgulati** [comment]
@tvernum I am having a problem in removing previous commits of mine from parameters branch. can you help me how can I split it?

**Comment by ojasgulati** [comment]
@tvernum I figured it myself. Force pushed, hope that's not a problem. 😃

**Comment by tvernum** [comment]
@elasticmachine Test this please

**Comment by ojasgulati** [comment]
@tvernum can you help me out why some tests are failing?

**Comment by gwbrown** [comment]
Hi @ojasgulati, thanks for taking this on!

It looks like the tests are failing because of an easy mistake to make that's why https://github.com/elastic/elasticsearch/issues/39666 was created: When you do this:
`RequestConverters.Params parameters = new RequestConverters.Params(request);`

Any writes to `parameters` are written through to the request.  With your changes, when you call `request.setParameters(parameters.getParam())`, the parameters that have already been written to `request` are retrieved from `request` and written again to `request` (which errors out because those parameters have already been set.

In order to address this issue, you'll need to refactor `RequestConverters.Params` so that its constructor does not take a `Request`, but instead takes no parameters and holds the parameters in a data structure (probably a `HashMap` or similar) in the `Params` object rather than writing to and retrieving them from a `Request`.

Does that make sense? You'll also need to merge `master` into your branch as there have been a number of changes in the last week - they shouldn't conflict with these changes, but the CI tests won't pass without them.

**Comment by ojasgulati** [comment]
@gwbrown Thank you for your response that was very helpful. I will try to make changes as you mentioned.  😃

**Comment by ojasgulati** [comment]
@gwbrown I made changes. please review this

**Comment by gwbrown** [comment]
@elasticmachine test this please

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java
Comment (gwbrown) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: `.*` imports aren't allowed in Elasticsearch code, which is why the test run failed - I suspect this is a change your IDE made automatically.

You can find guidance on how to configure your IDE [in our Contributing guide](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md) in the "Java Language Formatting Guidelines" section.

**Comment by ojasgulati** [comment]
@gwbrown I didn't notice that. I removed wild card imports. Please review this 
Thanks 😃

**Comment by gwbrown** [comment]
Thanks! I'll take a look today. In the meantime,
@elasticmachine test this please

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java
Comment (gwbrown) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: It would be better for this name to be something like `asMap()` - just from the name, I would assume `getParam` is the reverse of `putParam`, but that's not the case.

**Review by AthenaEryma — CHANGES_REQUESTED** [review]
Thanks for your hard work on this PR! This is looking pretty good, with only a few pretty minor things left to address.

You'll also need to merge `master` into your branch as we [made a new release today](https://www.elastic.co/blog/elastic-stack-7-1-1-released) and due to how our build tests backwards comparability, you'll need to have the changes from after that release was made.

Finally, you can run the unit tests for the request converters locally with `./gradlew :client:rest-high-level:test` - it should run pretty quick, about 2 minutes on my machine - and it'll help make the feedback loop shorter than waiting for our CI infrastructure to run the whole test suite.

**Comment by ojasgulati** [comment]
@gwbrown Thank you for your time. I made changes as mentioned. Also, I ran ```./gradlew :client:rest-high-level:test``` and every test case is passing. 
Cheers 😃

**Comment by gwbrown** [comment]
Thank you!
@elasticmachine test this please

**Comment by gwbrown** [comment]
It looks like there was a file that was accidentally committed, which I removed in https://github.com/elastic/elasticsearch/pull/42128/commits/b7c4c195e587b9676c088560aeb3a99743c6908a as this is otherwise ready to merge once CI passes (the previous failure was due to https://github.com/elastic/elasticsearch/issues/42679 which should now be resolved).

@elasticmachine test this please

**Review by AthenaEryma — APPROVED** [review]
LGTM, will merge once CI passes.

**Comment by gwbrown** [comment]
Thanks again for your contribution @ojasgulati!

**Comment by ojasgulati** [comment]
@gwbrown Thanks for your help. I want to contribute more but the project is kinda big to me and difficult to understand at once. Could you please suggest me something how should I get started to this project.  Also, is there any chat or a group so that I could ask about my problems?

**Comment by gwbrown** [comment]
You're welcome! Elasticsearch is indeed a very big project and even most of us who work on it all day every day don't understand all of it at once. I recommend choosing an area that's of interest to you and focusing on that - we categorize issues by functional area (it's the green labels that start with `:`, `:Core/Features/Java High Level REST Client` in this case) each of which will all be more or less related to the same code.

The best place to ask for general help is going to be either [the Discuss forums](https://discuss.elastic.co/c/elasticsearch) (though that tends to be more about using Elasticsearch than developing it) or on the PR/issue you're working on.

If you're interested in working on Elasticsearch full-time feel free to [apply to one of our open positions](https://jobs.elastic.co/jobs/department/engineering#/).

**Comment by ojasgulati** [comment]
@gwbrown Thank you for your response. I will think about the area and try to write more test for that. And thank you for the job offer but I am a 3-rd year student 😅.

**Pull request merged**
